test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,394 @@
+#!/bin/bash
+
+##
+## Case Name: check-volume-levels
+## Preconditions:
+##    topology is nocodec with loopack in PCM0P -> PCM0C
+##    aplay should work
+##    arecord should work
+##    PCM0C capture PGA supports -50 - +30 dB range
+##    PCM0C capture PGA supports mute switch
+## Description:
+##    Set volume and mute switch to various values, measure
+##    volume gain from the actual levels and compare to set gains.
+## Case step:
+##    1. Start aplay
+##    2a. Capture 1st wav file
+##    2b. Capture 2nd wav file
+##    2c. Capture 3rd wav file
+##    3. Measure volume gains
+## Expect result:
+##    command line check with $? without error
+##
+
+source "$(dirname "${BASH_SOURCE[0]}")/../case-lib/lib.sh"
+
+TPLGREADER=../tools/sof-tplgreader.py
+PCM_ID=0
+CAP_FORMAT="S16_LE"
+CAP_RATE="48000"
+VOL_P30DB=80
+VOL_P10DB=60
+VOL_0DB=50
+VOL_M10DB=40
+VOL_M20DB=30
+VOL_M30DB=20
+VOL_MAX=$VOL_P30DB
+VOL_MIN=1
+VOL_MUTE=0
+VOL_NOM=$VOL_0DB
+APLAY_WAV=$(mktemp --suffix=.wav)
+ARECORD_WAV1=$(mktemp --suffix=.wav)
+ARECORD_WAV2=$(mktemp --suffix=.wav)
+ARECORD_WAV3=$(mktemp --suffix=.wav)
+
+#
+# Test volume and mute switch controls
+#
+
+#  0...1s channels 1-4 max gain
+#  1...2s channels 1-4 different gains
+#  2...3s channels 1-4 nominal gain
+#  3...4s channels 1-4 muted
+#  4...5s channels 1-4 nominal gain
+#  5...7s channels 1-4 muted
+#  7...8s channels 1-4 max gain
+#  8...9s channels 1-4 muted gain
+#  9..10s channels 1-4 min gain
+# 10..11s channels 1-4 different gains
+run_test_1 () {
+    cset_max
+    cset_unmute
+    wait
+    arecord_in_background "$ARECORD_WAV1" "$CAP_CHANNELS" 11
+    PID=$?
+    wait
+
+    cset_volume_diff1 "$CAP_CHANNELS"
+    wait
+
+    cset_nom
+    wait
+
+    cset_mute
+    wait
+
+    cset_unmute
+    wait
+
+    cset_mute
+    wait
+
+    cset_max
+    wait
+
+    cset_unmute
+    wait
+
+    cset_mutevol
+    wait
+
+    cset_min
+    wait
+
+    cset_volume_diff2 "$CAP_CHANNELS"
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 previous gains
+# 1..2s channels 1-4 different mute switches
+# 2..3s channels 1-4 all muted
+run_test_2 () {
+    arecord_in_background "$ARECORD_WAV2" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
+
+    cset_nom
+    cset_mute_diff1 "$CAP_CHANNELS"
+    wait
+
+    cset_mute
+    wait
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 muted
+# 1..2s channels 1-4  nominal gain
+# 2..3s different mute switches
+# 3..4s channels 1-4 nominal gain
+run_test_3 () {
+    arecord_in_background "$ARECORD_WAV3" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
```

Comment:
`wait` waits for all processes so the purpose of `PID` is not obvious here.

Reply:
I learned "wait" was a bad name for my own function. It's now called "pause ()". It uses "sleep 1" but if I get stable results with shorter level measurements I may shorten it for better test speed.

Reply:
Rename done.